services/src/pro/users/oidc.rs

```diff
@@ -0,0 +1,919 @@
+                reason: "Failed to verify claims".to_string()
+            })?;
+
+        if let Some(expected_access_token_hash) = claims.access_token_hash() {
+            let actual_access_token_hash = AccessTokenHash::from_token(
+                token_response.access_token(),
+                &id_token.signing_alg()
+                    .map_err(|_signing_error| OIDCLoginFailed { reason: "Unsupported Signing Algorithm".to_string()})?
+            ).map_err(|_signing_error| OIDCLoginFailed {
+                reason: "Unsupported Signing Algorithm".to_string()
+            })?;
+            if actual_access_token_hash != *expected_access_token_hash {
+                return Err(OIDCLoginFailed {
+                    reason: "Illegal access token".to_string()
+                });
+            }
+        }
+
+        //TODO: Failing at missing claims seems unnecessary for external users, but UI expects email and real name.
+        let user = ExternalUserClaims {
+            external_id: claims.subject().clone(),
+            email: match claims.email() {
+                None => {Err(OIDCLoginFailed { reason: "Missing e-mail claim".to_string()})}
+                Some(x) => {Ok(x.to_string())}
+            }?,
+            real_name: match claims.name() {
+                None => {Err(OIDCLoginFailed { reason: "Missing name claim".to_string()})}
+                Some(x) => {Ok(x.get(None).unwrap().to_string())} //TODO: There is no Local logic.
+            }?
+        };
+
+        let validity = match token_response.expires_in(){
+            None => {Err(OIDCLoginFailed { reason: "No Duration in token".to_string() })}
+            Some(x) => {Ok(x)}
+        }?;
+
+        Ok((user, Duration::milliseconds(validity.as_millis() as i64))) //TODO: Is that cast ok, how does Geo Engine handle conversions between durations?
+    }
+
+    //TODO: Is there a better way to do that for testing? Seems hacky.
+    pub fn from_oidc_with_static_tokens(value: Oidc) -> Self {
+        OIDCRequestsDB {
+            issuer: value.issuer.to_string(),
+            client_id: value.client_id.to_string(),
+            client_secret: value.client_secret.to_string(),
+            redirect_uri: value.redirect_uri.to_string(),
+            scopes: value.scopes,
+            users: Arc::new(Default::default()),
+            state_function: || CsrfToken::new(SINGLE_STATE.to_string()),
+            nonce_function: || Nonce::new(SINGLE_NONCE.to_string())
+        }
+    }
+}
+
+impl Default for OIDCRequestsDB {
+
+    fn default() -> Self {
+        OIDCRequestsDB {
+            issuer: "".to_string(),
+            client_id: "".to_string(),
+            client_secret: "".to_string(),
+            redirect_uri: "".to_string(),
+            scopes: vec![],
+            users: Arc::new(Default::default()),
+            state_function: CsrfToken::new_random,
+            nonce_function: Nonce::new_random
+        }
+    }
```

Comment:
No. I don't see a reasonable default to be honest. I did not want to wrap it into an Option at the ProContext (so that other tests can still work fine), but I guess that was too lazy and I need to do so.

Reply:
Okay, we should define `Default` only if we can really use it.
For tests, you could implement TestDefault.

Reply:
Removed the default, put OIDCRequestDB as Option with a try_from here: 477a5228523df29c3dd2ee0bb48d78cfa4337742 (fixed a bug here: 2ed0b38b9f66fcd782833ee250604dcd94de703f).